Add custom repos to /etc/yum.repos.d

kingsleyzissou opened the pull request:
This pull request includes:
The ability to specify repo customizations so that custom repositories will be saved to
the `/etc/yum.repos.d` directory in the built image. 

- [ ] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`

Information regarding our GitLab pipeline (Schutzbot):

CI will not be ran automatically if WIP label is applied or the PR is in DRAFT state, instead only
a link will be provided to the pipeline which can then be triggered manually if desired. To run the
CI automatically either switch the PR to ready or apply WIP+test label.

Outside contributors need manual approval from one of the osbuild-composer maintainers.

Schutzbot will only be triggered if all Tests jobs in GitHub workflow succeed.
-->

thozza reviewed (commented):
Nice work. I added a few comments 😇 

I think that we can introduce this feature in steps:

1. Allow configuration of YUM repositories which will be baed-into the image, but not used for the image content. This can be workarounded on image-builder side, because it can implement a logic to duplicate the user-provided repos as customizations, in addition to providing them as payload repositories if the user chooses to have this.
2. We can later make the repository to be (optionally?) used also for image content.

In such a case, it is reasonable to let the user to provide the GPG key in the Repo definition as URL first. If we would then like to use the repo for image content, composer could fetch it once and embed it in the manifest for the purpose of the `rpm` stage. Of course only if the repository has `gpgcheck` enabled.

I'm currently working on the "custom files in /etc" feature. Once it is merged, we should have internal tool-agnostic representation of File and Directory with helper functions to generate the needed osbuild stages to create them in the image. It should be then fairly easy to use the same internal functions (that would be used for custom files in /etc) internally to save the user-repo's GPG key in the image and use it with `file:///...` URL in the YUM repo configuration.

IMO this can be introduced in steps and in backward compatible way (e.g. composer could determine if the GPG is provided in the customization "literally" or as an URL and handle it differently).

achilleas-k commented:
Regarding the GPG key issue: I had a little one-on-one with @kingsleyzissou on Friday to figure out some things about this together and it turns out we can import the keys into the database without needing to write files.  The `org.osbuild.rpm` stage imports keys listed in the options before installing packages.  If we have a key for a repository that will be configured / baked into the image, we can put the key in the OS pipeline's rpm stage options and it will be imported, even if the key isn't used for any of the packages that will be installed at that stage.

kingsleyzissou commented:
Thanks for updating this PR to add a comment @achilleas-k 🙌. Was going to get to that this morning but you beat me to it

thozza commented:
> Regarding the GPG key issue: I had a little one-on-one with @kingsleyzissou on Friday to figure out some things about this together and it turns out we can import the keys into the database without needing to write files. The `org.osbuild.rpm` stage imports keys listed in the options before installing packages. If we have a key for a repository that will be configured / baked into the image, we can put the key in the OS pipeline's rpm stage options and it will be imported, even if the key isn't used for any of the packages that will be installed at that stage.
> 
> If the repository is used for packages in the OS rpm stage, the key will be imported anyway.

I'm not a big fan of doing something like this TBH, at least for the following reasons:

- It is not obvious from a repo definition which GPG key it uses and where it can be inspected.
- If the repo is not really used for any packages baked into the image (which is IMO a reasonable use case), then importing the key seems counter-intuitive. Sure, the user will be asked to import it once DNF would need to use it for the first time, but that's OK, because the user gets a chance to verify the key (not that anyone does that).
- If a repository in `/etc/yum.repos.d` has `gpgcheck` enabled, then I think it is reasonable to expect that the GPG key to be used is specified in the repo definition.

kingsleyzissou commented on internal/blueprint/customizations.go:
```diff
@@ -117,6 +118,24 @@ type OpenSCAPCustomization struct {
+	Id            string   `json:"id" toml:"id"`
+	BaseURL       []string `json:"baseurl,omitempty" toml:"baseurl,omitempty"`
+	GPGKey        []string `json:"gpgkey,omitempty" toml:"gpgkey,omitempty"`
+	Cost          *int     `json:"cost,omitempty" toml:"cost,omitempty"`
```
Yeah cool, I'll update the PR with that.

kingsleyzissou commented on internal/blueprint/customizations.go:
```diff
@@ -117,6 +118,24 @@ type OpenSCAPCustomization struct {
+	BaseURL       []string `json:"baseurl,omitempty" toml:"baseurl,omitempty"`
+	GPGKey        []string `json:"gpgkey,omitempty" toml:"gpgkey,omitempty"`
+	Cost          *int     `json:"cost,omitempty" toml:"cost,omitempty"`
+	PayloadGPGKey *string  `json:"payload_gpgkey,omitempty" toml:"payload_gpgkey,omitempty"`
```
So this was regarding the issue I was having with the gpgkeys. If we want to add the package to the image at image-build we need to import the gpgkey either with the key or a file pointing to the key. The yum repo file doesn't accept a gpgkey though, so in that case we would need a url to a gpgkey or a file uri. I'm not a fan, but this was a way to provide both in the customizations as part of the WIP.

kingsleyzissou replied:
Ahhh... nevermind this previous comment. I missed your comment above about doing this in stages and letting image-builder handle the payload repos. Will fix this up

kingsleyzissou commented on internal/osbuild/yum_repos_stage.go:
```diff
@@ -23,6 +25,32 @@ type YumRepository struct {
+		RepoGPGCheck: repo.RepoGPGCheck,
+	}
+
+	if repo.Name != nil {
+		yr.Name = *repo.Name
+	}
+
+	if repo.Mirrorlist != nil {
+		yr.Mirrorlist = *repo.Mirrorlist
+	}
+
+	if repo.Metalink != nil {
+		yr.Metalink = *repo.Metalink
+	}
```
Okay cool. Makes sense

kingsleyzissou commented on internal/osbuild/yum_repos_stage.go:
```diff
@@ -23,6 +25,32 @@ type YumRepository struct {
 	RepoGPGCheck   *bool    `json:"repo_gpgcheck,omitempty"`
 }
 
+func NewYumRepository(repo blueprint.RepositoryCustomization) *YumRepository {
```
This is something Achilleas and I spoke about on Friday and it totally makes sense. We would have to change the `rpmmd.Repoconfig` to have a slice of `basurl` and `gpgkey`... or stick with a string for `baseurl` and `gpgkey` in the customizations.

achilleas-k commented:
We do this anyway when a user specifies a repo for a payload package.  We import the key into the rpm db but don't have a repo configured for it.  Is it regarded good practice to have a gpg key for each configured repo in `/etc/pki/...`?  If it is, I see your point.

thozza commented:
> We do this anyway when a user specifies a repo for a payload package. We import the key into the rpm db but don't have a repo configured for it. Is it regarded good practice to have a gpg key for each configured repo in `/etc/pki/...`? If it is, I see your point.

Fair point with the user-repos 🤔 I'm not sure how to determine if it is a good practice and not just my opinion... It just feels weird if a repo definition would specify `gpgcheck` enabled and not have any `gpgkey` specified. So if we can make it so that this does not happen, I would prefer it. I'll probably reach out to the DNF/RPM team and ask them...

achilleas-k commented:
> I'll probably reach out to the DNF/RPM team and ask them...

Good idea :+1:

kingsleyzissou commented:
Still to do: GCE YUMRepos. I can start that while we wait for feedback from the dnf team.

I'm piggybacking of the changes @bcl is making in #3230 (I'll drop those commits once the PR has landed)

kingsleyzissou commented on internal/distro/image_config.go:
```diff
@@ -34,6 +35,7 @@ type ImageConfig struct {
 	// Disable documentation
 	ExcludeDocs *bool
 
+	YUMRepos []rpmmd.RepoConfig
```
Going to change this to a map so we can still accept a filename

kingsleyzissou commented on internal/manifest/os.go:
```diff
@@ -74,8 +79,9 @@ type OSCustomizations struct {
-	Users  []users.User
+	Groups   []users.Group
+	Users    []users.User
+	YUMRepos []rpmmd.RepoConfig
```
Same here. Going to change this to a map to accept filename

thozza replied:
Alternatively you can define it as a new type in `rpmmd`, which would be IMO also reasonable. Something like:
```go
type YUMRepoFile map[string]RepoConfig
```

kingsleyzissou commented on internal/rpmmd/repository.go:
```diff
@@ -375,3 +378,45 @@ func (packages PackageList) ToPackageInfos() []PackageInfo {
+	type compatType struct {
+		aliasType
+
+		BaseURL string `json:"Baseurl,omitempty"`
```
Repo config doesn't actually have any json tags... so I did this for now. Any suggestions welcome.

bcl replied:
I'm fairly sure `RepoConfig` never gets serialized to/from json. It's an internal struct that is filled in manually by `loadRepositoriesFromFile` which uses `repository` as the on-disk json struct.

kingsleyzissou replied:
Ah right, I had to add this in though because the `regression-old-worker` test kept failing with an rpmmd error because it couldn't unmarshal the fields. 

Just to check, is it not getting marshalled here though?
https://github.com/osbuild/osbuild-composer/blob/da2e80f99f65bf708e0c7c3a31ee4844846287d6/internal/worker/json.go#L224

bcl replied:
Ah, yes it would appear so :/ The function says it is temporary, so maybe that will change. But I really have no idea, not having touched the worked code at all.

achilleas-k replied:
It's probably time to drop the `compatJob` part here.  The change it was covering for was made almost a year ago (94c7fda779).  Since da2e80f99f we only care about and test workers from 3 versions back, so I think we're safe to remove it and it's not completely out of scope for this PR I think.

achilleas-k replied:
Looking at it again, it seems that even in the new format we would serialise `rpmmd.RepoConfig` as part of the `rpmmd.PackageSet` in `DepsolveJob`.  That said, we've been serialising `rpmmd.RepoConfig` for ages, since the inception of the `DepsolveJob` 0f90aa9c782cace1fc88c39c7ec752bd1f12d799, so maybe it's fine?

kingsleyzissou replied:
Because we're simplifying this PR, this technically could be done in a separate PR? Happy to add the json tags and create a PR for that, wdyt?

thozza replied:
Whatever is easier for you...

kingsleyzissou replied:
Did it here :)

kingsleyzissou commented on internal/distro/fedora/distro.go:
```diff
@@ -767,6 +767,8 @@ func (t *imageType) Manifest(customizations *blueprint.Customizations,
 		mergedRepos = append(mergedRepos, repo)
 	}
 
+	mergedRepos = append(mergedRepos, customRepos...)
```
Wasn't too sure how to work around this... this seems a bit meh. But the custom repos get removed otherwise.

thozza commented on internal/distro/fedora/distro.go:
```diff
@@ -856,6 +858,12 @@ func (t *imageType) checkOptions(customizations *blueprint.Customizations, optio
 	}
 
+	repos := customizations.GetRepositories()
+	err = blueprint.ValidateCustomRepositories(repos)
```
I think that it would be better to validate when we unmarshal from TOML, but this is also OK.

thozza commented on internal/distro/fedora/images.go:
```diff
@@ -48,6 +49,7 @@ func osCustomizations(
 	osc.ExtraBaseRepos = osPackageSet.Repositories
 
 	osc.Containers = containers
+	osc.CustomRepos = customRepos
```
You should take this from passed blueprint customizations instead, not as a separate function argument.

thozza reviewed (changes requested):
I added many comments, sorry. I'll need to do another iteration, but I'm submitting what I have now...

kingsleyzissou commented:
So to briefly summarise our call, this is what we discussed:
- only save repos to `/etc/yum.repos.d` and not use the custom repos as payload repos (this will greatly simplify the pr)
- resolve url gpgkeys and import them, but keep the url in the final `/etc/yum.repos/d` repo file
- save inline gpgkeys to `/etc/pki` and replace them in the `/etc/yum.repos.d/` with a `file://` uri
- either use a map or specify the filename of the custom repository in the customizations rather than saving all the repos to one file.

kingsleyzissou commented on internal/rpmmd/repository.go:
```diff
@@ -423,3 +424,35 @@ func (r *RepoConfig) UnmarshalJSON(data []byte) error {
 	return nil
 }
+
+func RepoConfigFromBP(customRepos []blueprint.RepositoryCustomization) []RepoConfig {
+	var repos []RepoConfig
+	for _, customRepo := range customRepos {
+		repos = append(repos, convertCustomRepo(customRepo))
+	}
+	return repos
+}
+
+func convertCustomRepo(customRepo blueprint.RepositoryCustomization) RepoConfig {
+	urls := []string{}
+	urls = append(urls, customRepo.BaseURLs...)
+
+	keys := []string{}
+	keys = append(keys, customRepo.GPGKeys...)
+
+	repoConfig := RepoConfig{
+		Id:           customRepo.Id,
+		BaseURLs:     urls,
+		GPGKeys:      keys,
+		Name:         customRepo.Name,
+		Metalink:     customRepo.Metalink,
+		Mirrorlist:   customRepo.Mirrorlist,
+		GPGCheck:     customRepo.GPGCheck,
+		RepoGPGCheck: customRepo.RepoGPGCheck,
+		Priority:     customRepo.Priority,
+		Enabled:      customRepo.Enabled,
+		IgnoreSSL:    !customRepo.SSLVerify,
+	}
+
+	return repoConfig
+}
```
Okay cool, will move that. That explanation makes sense and hadn't crossed my mind

kingsleyzissou commented on internal/distro/distro.go:
```diff
@@ -121,7 +121,7 @@ type ImageType interface {
 	// deprecation notices) generated by the manifest.
 	// The packageSpecSets must be labelled in the same way as the originating PackageSets.
-	Manifest(b *blueprint.Customizations, options ImageOptions, repos []rpmmd.RepoConfig, packageSpecSets map[string][]rpmmd.PackageSpec, containers []container.Spec, seed int64) (Manifest, []string, error)
+	Manifest(b *blueprint.Customizations, options ImageOptions, repos []rpmmd.RepoConfig, packageSpecSets map[string][]rpmmd.PackageSpec, containers []container.Spec, customRepos []rpmmd.RepoConfig, seed int64) (Manifest, []string, error)
```
I think this was because I was trying to do some cleverness with the file resolver and passing the keys down, but I don't think there'll be any need for this with the simplifications... so I'll revert back to using the customizations.

kingsleyzissou commented on internal/osbuild/rpm_key_input.go:
```diff
@@ -0,0 +1,19 @@
+	"fmt"
+)
+
+type RPMKeyInlineInput struct {
+	InlineFile *FilesInput `json:"inlinefile"`
+}
+
+func (RPMKeyInlineInput) isStageInputs() {}
+
+func NewRPMInlineInput(contents []byte) Inputs {
+	input := NewFilesInput(NewFilesInputSourcePlainRef([]string{
+		fmt.Sprintf("%x", sha256.Sum256(contents)),
+	}))
+	return &RPMKeyInlineInput{InlineFile: input}
+}
```
Cool, yeah that makes sense. Like I said in the call, might pick your brains on this when I get to it.

thozza commented:
> So to briefly summarise our call, this is what we discussed:
> 
> * only save repos to `/etc/yum.repos.d` and not use the custom repos as payload repos (this will greatly simplify the pr)
> * resolve url gpgkeys and import them, but keep the url in the final `/etc/yum.repos/d` repo file

TBH, I don't this this ^^^ is necessary at all if we are going to not use the custom repository for installing content in the image. DNF will import the key when installing the first package from the custom repo. So we may as well skip importing it at all.

> * save inline gpgkeys to `/etc/pki` and replace them in the `/etc/yum.repos.d/` with a `file://` uri
> * either use a map or specify the filename of the custom repository in the customizations rather than saving all the repos to one file.

kingsleyzissou commented on internal/osbuild/yum_repos_stage.go:
```diff
@@ -68,13 +71,45 @@ type YumReposStageOptions struct {
+		Metalink:     repo.Metalink,
+		BaseURL:      urls,
+		GPGKey:       keys,
+		GPGCheck:     &repo.GPGCheck,
+		RepoGPGCheck: &repo.RepoGPGCheck,
+		Enabled:      &repo.Enabled,
+		Priority:     &repo.Priority,
```
I managed to work around this when deserialising the json/toml and only explicitly setting the fields that are there.
The code is slightly more verbose there for the unmarshaling functions, but it means we don't have to change those values to pointers in `rpmmd.RepoConfig`. Lmk what you think

kingsleyzissou commented on internal/rpmmd/repository.go:
```diff
@@ -375,3 +378,45 @@ func (packages PackageList) ToPackageInfos() []PackageInfo {
+	}
+
+	if compatRepo.BaseURL != "" {
+		compatRepo.BaseURLs = strings.Split(compatRepo.BaseURL, ",")
```
Ah  I missed this, I've left it as a list since the `yum_repo_stage` takes as a list of baseurls as an option.
Happy to change that though. But not sure if it makes a huge difference.

kingsleyzissou commented on test/data/manifests/rhel_7-x86_64-qcow2_customize-boot.json:
```diff
@@ -146,6 +146,22 @@
             "user": 0,
             "group": 0
           }
+        ],
```
Ah I think I missed a step for the rhel implementation. The yum repo stage is not being added to the generated manifest.

kingsleyzissou replied:
The `imageConfig.YUMRepos` were overwriting the `osc.YUMRepos`

kingsleyzissou commented on test/data/manifests/centos_8-aarch64-qcow2_customize-boot.json:
```diff
@@ -131,6 +131,22 @@
             "group": 0
           }
+        ],
+        "repositories": [
```
Same as above

kingsleyzissou commented on test/data/manifests/centos_9-aarch64-qcow2_customize-boot.json:
```diff
@@ -131,6 +131,22 @@
             "group": 0
           }
+        ],
+        "repositories": [
```
Same as above

thozza commented on internal/rpmmd/test/repository_test.go:
```diff
@@ -228,14 +228,14 @@ func TestOldWorkerRepositoryCompatMarshal(t *testing.T) {
 	}{
 		{
-			repoJSON: []byte(`{"Name":"fedora","BaseURLs":["http://example.com/fedora"],"Metalink":"","Mirrorlist":"","GPGKeys":null,"GPGCheck":false,"RepoGPGCheck":false,"IgnoreSSL":false,"MetadataExpire":"","RHSM":false,"ImageTypeTags":null,"PackageSets":null,"Baseurl":"http://example.com/fedora"}`),
+			repoJSON: []byte(`{"Name":"fedora","BaseURLs":["http://example.com/fedora"],"Metalink":"","Mirrorlist":"","GPGKeys":null,"GPGCheck":null,"RepoGPGCheck":null,"IgnoreSSL":false,"MetadataExpire":"","RHSM":false,"ImageTypeTags":null,"PackageSets":null,"Baseurl":"http://example.com/fedora"}`),
```
nitpick: seeing the `null` in the JSON string, I think that it would be good to eventually use dedicated structures for marshaling or at least add JSON tags to the `RepoConfig` struct and make these `omitempty`. However not a blocker for this PR from my PoV.

kingsleyzissou replied:
Oh I did add JSON tags to RepoConfig, but I did it in a later commit... thinking about this further, I probably should have done that first.

thozza replied:
Sorry, I was reviewing the PR commit by commit and thinking about this now, you are right that you added tags in later commit, but I didn't get back to my comment. Adding them first would be probably better.

kingsleyzissou replied:
Yeah no, I figured you did them commit by commit... I've added the json tags now in the commit you've commented on.

kingsleyzissou replied:
Yeah no, I figured you did them commit by commit... I've added the json tags now in the commit you've commented on.

thozza commented on internal/store/store_test.go:
```diff
@@ -444,6 +446,7 @@ func (suite *storeTest) TestNewSourceConfigWithMirrorlist() {
 	myRepoConfig := rpmmd.RepoConfig{
 		Name:       "testRepo",
 		Mirrorlist: "testURL",
+		IgnoreSSL:  false,
```
I'm wondering why was this needed, since in this commit (https://github.com/osbuild/osbuild-composer/pull/3228/commits/d34fe71bdded6dd6f3a46c8111ba6da0bb375062) I don't see any change related to `IgnoreSSL` struct member 🤔

kingsleyzissou replied:
Good catch! I think I made `IgnoreSSL` a pointer and then undid it - but that's my only guess.

thozza commented on internal/blueprint/repository_customizations.go:
```diff
@@ -326,3 +330,84 @@ func validateCustomRepository(repo *RepositoryCustomization) error {
+
+	for _, repo := range repos {
+		convertedRepo := repo.customRepoToRepoConfig()
+		m[repo.Filename] = append(m[repo.Filename], convertedRepo)
```
If I read the implementation correctly, `repo.Filename` may be an empty string. It could make sense to add a method for the `RepositoryCustomization` struct which will generate one if not set explicitly by user. We discussed that we could default to the `<repo_id>.repo`.

kingsleyzissou replied:
So at the moment, I'm setting the filename in the unmarshal functions if nothing is set - it might not the best place to do it and I could make that more explicit, like you suggest.

For reference:
https://github.com/osbuild/osbuild-composer/blob/fd61e02f2a3a86117a9147a2dd59040d06a3ed60/internal/blueprint/repository_customizations.go#L195

thozza replied:
Ah, I indeed missed that. I think that doing this in the custom unmarshal function may not be the best place. Specifically, because this will solve the problem of filling the not provided filename in CloudAPI where we just copy customizations from the compose request to the blueprint created on the fly https://github.com/osbuild/osbuild-composer/blob/03b093acb024675eab75b7fe266b3e6db2504dc7/internal/cloudapi/v2/handler.go#L142

We do not unmarshal there into the BP customizations so the missing filename would not be filled in in such case.

WRT the filename if it is not provided, I'll let others to comment as well, but I find a bit nicer if we simply use the repo ID as the repo filename.

In addition, the repo IDs must be unique across all configuration files, which is not currently enforced in any way and it would probably make sense to check as it would solve some corner cases. To complicate things more, the user-provided repo ids can conflict with repos specified in the image configuration (currently used only by GCE images).

kingsleyzissou replied:
Yeah no problem at all... I like your solution better, it's simpler and more elegant and avoids magic. I've fixed that already and will push the changes in a bit :). 

It's also less of headache if we just save them to a repo with their own id rather than grouping all the repos with the same filename together.

thozza commented on internal/blueprint/repository_customizations.go:
```diff
@@ -326,3 +330,84 @@ func validateCustomRepository(repo *RepositoryCustomization) error {
+		return nil
+	}
+
+	m := make(map[string][]rpmmd.RepoConfig)
```
nitpick: you should probably pass `len(repos)` as the initial capacity of the map.

thozza commented on internal/blueprint/repository_customizations.go:
```diff
@@ -326,3 +330,84 @@ func validateCustomRepository(repo *RepositoryCustomization) error {
+// DeepCopy returns a deep copy of the given slice of RepositoryCustomization
+// Since we are replacing the gpgkeys of the customization with the file path
+// we don't want to modify the original slice.
+func DeepCopyCustomRepositories(repos []RepositoryCustomization) []RepositoryCustomization {
```
I don't think this is needed. My suggestion would be to:

Merge `GetInlineGPGKeysFromCustomRepositories()` and `RepoCustomizationsToRepoConfig()` into a single function, which returns `(map[string][]rpmmd.RepoConfig, []*fsnode.File, error)`. In a single iteration, you create a new `rpmmd.RepoConfig` which already has deep copy of all values from the repository customization structure. You would then generate a `*fsnode.File` for the repo if the GPG key was provided inline and you would modify the `rpmmd.RepoConfig` to point to a file, before adding it to the map of slices of repo configs. This should not create any issues (AFAICT).

Doing the conversion to internal representations in two steps is the only reason that you need this helper function and it also creates possibility to call them in the wrong order and not have inline GPG keys converted to file paths. Modifying the `rpmmd.RepoConfig` after conversion in a single function will eliminate any need for deep copy since you won't ever modify the customization structure.

Also I would appreciate if you could add unit tests for all of these newly added helper functions.

kingsleyzissou replied:
Okay, cool sounds good to me. I will try merge the functions like you suggest and yes, absolutely, I'll add unit tests... just wanted to make sure I was on the right track before doing that :)

thozza replied:
You are definitely on the right track, you got it right! My suggestion would just simplify the implementation, but the resulting behavior should stay the same 😉

thozza commented on internal/distro/fedora/images.go:
```diff
@@ -143,6 +143,31 @@ func osCustomizations(
+
+	// since we are replacing the inline gpgkeys with their file paths,
+	// we need to make a copy of the repos so we don't modify the original
+	customRepos := blueprint.DeepCopyCustomRepositories(c.GetRepositories())
+
+	// get the inline gpg keys from the repos and convert them to fs node files
+	// so we can save them to disk. This step also swaps the inline gpg key
+	// with the path to the file in the os file tree
+	gpgKeyFiles, err := blueprint.GetInlineGPGKeysFromCustomRepositories(customRepos)
+	if err != nil {
+		panic(fmt.Sprintf("failed to convert inline gpgkeys to fs node files: %v", err))
+	}
```
As I commented elsewhere, I propose to get `gpgKeyFiles` and `yumRepos` in a single function call without the need to make a copy of the customizations.

thozza reviewed (commented):
Very nice iteration. Added some suggestions inline...

thozza commented:
@kingsleyzissou one thing I realized that's missing is the exposure of this customization through the CloudAPI. I guess that's still on your radar, but wanted to call it out.

kingsleyzissou commented:
> @kingsleyzissou one thing I realized that's missing is the exposure of this customization through the CloudAPI. I guess that's still on your radar, but wanted to call it out.

Yeah it is indeed, I kind of have a WIP that I turned into a debugging PR. I think I'll do the cloudapi stuff in a separate PR.

achilleas-k commented on internal/blueprint/repository_customizations.go:
```diff
@@ -0,0 +1,326 @@
+	Filename     string   `json:"filename,omitempty" toml:"filename,omitempty"`
+}
+
+func (rc *RepositoryCustomization) UnmarshalTOML(data interface{}) error {
```
Why do we need the custom unmarshaller?  Is it just for checking required fields and their types?  Would it be easier to check it when reading instead of needing to define our own functions?

kingsleyzissou replied:
I think this was a suggestion from @thozza... and then we're doing the validation inside of this function rather than doing it when generating the manifest.

thozza replied:
I suggested this. Doing it in e.g. `GetRepositories()` would be nice, but it does not return any error, but it would not be a problem to add one. The custom unmarshaller is a bit of an overkill, I admit.

kingsleyzissou replied:
Ah good point, I'll add it to the `GetRepositories()` function instead and return an error

kingsleyzissou replied:
Updated this... what I don't love about it though is that I have to call this twice,
Here (to validate the repos):
https://github.com/osbuild/osbuild-composer/blob/18240dee91343c50bacf2fb1dd102fc0924b54a2/internal/distro/fedora/distro.go#L849

And here:
https://github.com/osbuild/osbuild-composer/blob/18240dee91343c50bacf2fb1dd102fc0924b54a2/internal/distro/fedora/images.go#L147-L152

It feels weird not calling the validation earlier and waiting until the second call and panicking. Would it be better to export the validation function again at that point? Or is this approach okay?

thozza replied:
@kingsleyzissou I think that this approach is OK. As you noted in the comment, we should not reach the second call if there was an error and if we did, it is a programming error.

Going forward, it would probably make sense to add a `Validate()` method to the `blueprint.Customizations` structure and call only that from anywhere (in image definitions, APIs, etc). Such method should then validate each specific customization using its own validator.

achilleas-k commented on tools/test-case-generators/format-request-map.json:
```diff
@@ -943,6 +943,21 @@
+          ],
+          "repositories": [
+            {
+              "id": "webmin-noarch",
+              "name": "Webmin - noarch",
+              "baseurls": [
+                "https://download.webmin.com/download/yum"
+              ],
```
I'm not sure about adding a repository we don't control (or can rely on) as part of our tests.  I guess it's fine since they're not used during the build, but maybe we can add one of our own (maybe something like `sap-rhui-azure` from our snapshots) or even a fake one (like example.org)?

achilleas-k reviewed (commented):
Nice work!  Thanks for working on this for so long through all the changes :)
I have some comments below but they're mostly nitpicks and questions.

thozza commented on internal/blueprint/repository_customizations.go:
```diff
@@ -324,3 +327,73 @@ func validateCustomRepository(repo *RepositoryCustomization) error {
+	}
+
+	repoMap := make(map[string][]rpmmd.RepoConfig, len(repos))
+	gpgKeyFiles := []*fsnode.File{}
```
IMO it would be better to declare the variable without assignment (so `var gpgKeyFiles []*fsnode.File`, so that if there are no Files for GPG keys after the conversion, the returned value will be `nil`. Checking the `len()` of such value will work fine, but you'll need to adjust unit tests.

thozza commented on internal/blueprint/repository_customizations_test.go:
```diff
@@ -93,3 +96,340 @@ func TestCustomRepoSerialization(t *testing.T) {
+	}
+}
+
+func TestRepoCustomizationsToGPGKeyFiles(t *testing.T) {
```
Functionally, this is all good. Logically, I think that you could test conversion of repo customizations to `RepoConfig` and `File` in a single test case. However I'm OK with keeping these test cases split.

thozza commented on test/cases/custom-repos.sh:
```diff
@@ -0,0 +1,198 @@
+REPO_ID="webmin-noarch"
+REPO_NAME="Webmin - noarch"
+REPO_FILENAME="custom.repo"
+REPO_BASEURL="https://download.webmin.com/download/yum"
```
As mentioned elsewhere, using `example.com` or some of our own repos would be better.

kingsleyzissou commented on internal/rpmmd/repository.go:
```diff
@@ -18,9 +18,9 @@ type repository struct {
+	Mirrorlist     string   `json:"mirrorlist,omitempty"`
 	GPGKey         string   `json:"gpgkey,omitempty"`
-	CheckGPG       bool     `json:"check_gpg,omitempty"`
+	GPGCheck       bool     `json:"check_gpg,omitempty"`
```
Yeah I've been wondering about that. At first it was super confusing for me. But I guess we convert them in different places... so maybe it is better to drop that commit and leave things as is.

kingsleyzissou commented on internal/rpmmd/repository.go:
```diff
@@ -375,3 +378,45 @@ func (packages PackageList) ToPackageInfos() []PackageInfo {
 }
+
+func (r RepoConfig) MarshalJSON() ([]byte, error) {
+	type aliasType RepoConfig
```
Good point, will add that.

kingsleyzissou commented on internal/blueprint/repository_customizations.go:
```diff
@@ -0,0 +1,326 @@
+	case interface{}:
+		urls, ok := baseurls.([]interface{})
+		if !ok {
+			return fmt.Errorf("TOML unmarshal: baseurl must be a string")
```
Think like you mention... I can simplify the json unmarshaller like you've done with the directory customization... this will get rid of that.

thozza replied:
But if you meant doing it in `GetRepositories()`, then yeah, that should be possible and simplify things a lot!

kingsleyzissou commented on internal/blueprint/repository_customizations.go:
```diff
@@ -0,0 +1,326 @@
+}
+
+func (rc *RepositoryCustomization) UnmarshalJSON(data []byte) error {
+	var dataMap map[string]interface{}
```
Cool cool, I'll update this and simplify.

thozza commented on .gitlab-ci.yml:
```diff
@@ -571,6 +571,11 @@ filesystem.sh:
   variables:
     SCRIPT: filesystem.sh
 
+custom-repos.sh:
+  extends: .integration
+  variables:
+    SCRIPT: custom-repos.sh
```
Keeping the test here for now is OK from my PoV. We should probably rethink how we test customizations on premise in general...

kingsleyzissou commented on internal/rpmmd/repository.go:
```diff
@@ -49,15 +49,21 @@ func (r *RepoConfig) Hash() string {
 		return fmt.Sprintf("%T", b)
 	}
+	bpts := func(b *bool) string {
+		if b == nil {
+			return fmt.Sprintf("%T", false)
+		}
```
Yeah I think the `bpts` function isn't needed anymore then.

kingsleyzissou commented on internal/blueprint/repository_customizations.go:
```diff
@@ -0,0 +1,326 @@
+	if repo.GPGCheck != nil && *repo.GPGCheck && len(repo.GPGKeys) == 0 {
+		return fmt.Errorf("Repository gpg check is set to true but no gpg keys are provided")
+	}
+	return nil
```
I'll add this in the cloudapi PR :)

thozza replied:
OK

thozza reviewed (approved):
Great work! 👏 

I think that everything has been sorted out and some suggestions will be implemented as a follow-up.

achilleas-k reviewed (approved):
Great work.  Thanks for all the effort you put in to this.
**LGTM!**

achilleas-k commented:
Wow, CI is very unhappy.  Some weird errors showing up.

kingsleyzissou commented:
@achilleas-k tx!
Yeah I just rebased... the previous iteration broke the api tests

It was changing the `bts` function inside the `rpmmd.Hash` function from `%T` to `%v`.
I've reverted that now (and tested) and CI _should_ go green now
